[AArch64] Replace expensive move from wzr by two moves via floating point immediate

juliannagele opened the pull request:
We've noticed that inserting 0 into a known vector lane is implemented via a move from wzr, i.e., moving between register banks.  We think it will be cheaper (and have seen improvements on our benchmarks) to materialize 0 into a floating point register and insert from there.

juliannagele committed:
[ISel] Replace expensive mov from wzr by two moves via fpr

davemgreen reviewed (commented):
I think in theory this shouldn't be much of an improvement, if the existing operation is already split and one of the mops is essentially the new instruction. In practice it is probably a good idea though.

There are certain cpus where it will not be helpful. Do you mind adding a tuning feature that keeps the old patterns for cortex-a53, a55, a510, a520 and a310. Those are all in-order and this will just be extra instructions.

juliannagele commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -8035,6 +8029,18 @@ def MOVIv2d_ns   : SIMDModifiedImmVectorNoShift<1, 1, 0, 0b1110, V128,
                                                 "movi", ".2d",
                    [(set (v2i64 V128:$Rd), (AArch64movi_edit imm0_255:$imm8))]>;
 
+def : Pat<(vector_insert (v8f16 V128:$Rn), (f16 fpimm0), (i64 VectorIndexH:$imm)),
+          (INSvi16lane V128:$Rn, VectorIndexH:$imm,
+          (v8f16 (MOVIv2d_ns (i32 0))), (i64 0))>;
+
+def : Pat<(vector_insert (v4f32 V128:$Rn), (f32 fpimm0), (i64 VectorIndexS:$imm)),
+          (INSvi32lane V128:$Rn, VectorIndexS:$imm,
+          (v4f32 (MOVIv2d_ns (i32 0))), (i64 0))>;
+
+def : Pat<(vector_insert (v2f64 V128:$Rn), (f64 fpimm0), (i64 VectorIndexD:$imm)),
+          (INSvi64lane V128:$Rn, VectorIndexD:$imm,
+          (v2f64 (MOVIv2d_ns (i32 0))), (i64 0))>;
```
Hm, sorry not sure I understand, without those patterns we get the move from wzr (which we want to avoid).

juliannagele replied:
Ohh, I see, it is, yes :)

davemgreen reviewed (commented):
> Sure, do you maybe have an example I can work off, not sure how to go about this.

I think that DisableFastIncVL/HasFastIncVL shows how to add a subtarget feature and make patterns dependant on it. The old patterns can be made dependant on it, I believe, and it can be added to the processors that would not benefit. (FastIncVL is perhaps a bit odd as it doesn't appear to be used anywhere).

juliannagele committed:
fixup! [ISel] Replace expensive mov from wzr by two moves via fpr

Instead of introducing new patterns, guard exisiting ones by tuning feature

juliannagele commented:
Thanks! And you were right about not needing new patterns, so I've removed the ones I added and put the existing `INS .. WZR` ones behind a feature and added that to the cpus you mentioned.

> > Sure, do you maybe have an example I can work off, not sure how to go about this.
> 
> I think that DisableFastIncVL/HasFastIncVL shows how to add a subtarget feature and make patterns dependant on it. The old patterns can be made dependant on it, I believe, and it can be added to the processors that would not benefit. (FastIncVL is perhaps a bit odd as it doesn't appear to be used anywhere).

juliannagele committed:
Merge branch 'main' into insertzero_lowering

juliannagele committed:
fixup! fixup! [ISel] Replace expensive mov from wzr by two moves via fpr

davemgreen reviewed (approved):
LGTM, thanks

fhahn committed:
Merge branch 'main' into insertzero_lowering

fhahn reviewed (approved):
LGTM, thanks

juliannagele commented:
/cherry-pick e333d60
